p2panda-rs/src/test_utils/db/domain.rs

```diff
@@ -0,0 +1,1246 @@
+
+            next_args.backlink = latest_entry.map(|entry| entry.hash());
+            next_args.skiplink = skiplink;
+            next_args.seq_num = seq_num;
+            next_args.log_id = log_id;
+        }
+    };
+
+    Ok(next_args)
+}
+
+/// Persist an entry and operation to storage after performing validation of claimed values against
+/// expected values retrieved from storage.
+///
+/// Returns the arguments required for constructing the next entry in a bamboo log for the
+/// specified author and document.
+///
+/// This method is intended to be used behind a public API and so we assume all passed values
+/// are in themselves valid.
+///
+/// # Steps and Validation Performed
+///
+/// Following is a list of the steps and validation checks that this method performs.
+///
+/// ## Validate Entry
+///
+/// Validate the values encoded on entry against what we expect based on our existing stored
+/// entries:
+///
+/// - Verify the claimed sequence number against the expected next sequence number for the author
+///     and log.
+/// - Get the expected backlink from storage.
+/// - Get the expected skiplink from storage.
+/// - Verify the bamboo entry (requires the expected backlink and skiplink to do this).
+///
+/// ## Ensure single node per author
+///
+/// - @TODO
+///
+/// ## Validate operation against it's claimed schema:
+///
+/// - @TODO
+///
+/// ## Determine document id
+///
+/// - If this is a create operation:
+///   - derive the document id from the entry hash.
+/// - In all other cases:
+///   - verify that all operations in previous_operations exist in the database,
+///   - verify that all operations in previous_operations are from the same document,
+///   - ensure that the document is not deleted.
+/// - Verify that the claimed log id matches the expected log id for this author and log.
+///
+/// ## Persist data
+///
+/// - If this is a new document:
+///   - Store the new log.
+/// - Store the entry.
+/// - Store the operation.
+///
+/// ## Compute and return next entry arguments
+pub async fn publish<S: StorageProvider>(
+    store: &S,
+    schema: &Schema,
+    encoded_entry: &EncodedEntry,
+    plain_operation: &PlainOperation,
+    encoded_operation: &EncodedOperation,
+) -> Result<EntryArgsResponse> {
+    //////////////////
+    // DECODE ENTRY //
+    //////////////////
+
+    let entry = decode_entry(encoded_entry)?;
+    let author = entry.public_key();
+    let log_id = entry.log_id();
+    let seq_num = entry.seq_num();
+
+    //////////////////////////////////
+    // VALIDATE ENTRY AND OPERATION //
+    //////////////////////////////////
+
+    // TODO: Check this validation flow is still correct.
+
+    // Verify that the claimed seq num matches the expected seq num for this author and log.
+    let latest_entry = store.get_latest_entry(author, log_id).await?;
+    let latest_seq_num = latest_entry.as_ref().map(|entry| entry.seq_num());
+    is_next_seq_num(latest_seq_num, seq_num)?;
```

Comment:
Interesting! This should be validated implicitly by `validate_log_integrity` inside of `validate_operation_with_entry` as the hash will otherwise be wrong of the backlink entry. I wonder if we can add a test for that just to be sure!

Reply:
Aha, yes, this is exactly what I was asking you to look at this morning :+1: these methods are exactly what we are using in `aquadgoog` too. I'll check this and remove if it's not needed.

Reply:
`aquadgoog`! amazing word

Reply:
I'll leave this here now as I think there may be more which we will catch in the `aquadoggo` pr, then I'll bring all that over here in one go.